mirgecom/diffusion.py

```diff
@@ -0,0 +1,116 @@
+r""":mod:`mirgecom.diffusion` computes the diffusion operator.
+
+.. autofunction:: diffusion_operator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import math
+import numpy as np
+import numpy.linalg as la  # noqa
+from pytools.obj_array import make_obj_array, obj_array_vectorize
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from meshmode.dof_array import thaw, DOFArray
+from grudge.symbolic.primitives import TracePair
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+
+
+def _q_flux(discr, alpha, u_tpair):
+    normal = thaw(u_tpair.int[0].array_context, discr.normal(u_tpair.dd))
+    flux_weak = math.sqrt(alpha)*u_tpair.avg*normal
+    return discr.project(u_tpair.dd, "all_faces", flux_weak)
+
+
+def _u_flux(discr, alpha, q_tpair):
+    normal = thaw(q_tpair.int[0].array_context, discr.normal(q_tpair.dd))
+    flux_weak = math.sqrt(alpha)*make_obj_array([np.dot(q_tpair.avg, normal)])
+    return discr.project(q_tpair.dd, "all_faces", flux_weak)
+
+
+def _operator(discr, alpha, w):
+    u = make_obj_array([w])
+
+    dir_u = discr.project("vol", BTAG_ALL, u)
+
+    q = discr.inverse_mass(
+        -math.sqrt(alpha)*discr.weak_grad(u[0])
+        +  # noqa: W504
+        discr.face_mass(
+            _q_flux(discr, alpha=alpha, u_tpair=interior_trace_pair(discr, u))
+            + _q_flux(discr, alpha=alpha,
+                u_tpair=TracePair(BTAG_ALL, interior=dir_u, exterior=-dir_u))
```

Comment:
It might be useful to allow more flexibility in boundary conditions, e.g. Dirichlet on this tag, Neumann on another.

Reply:
I have that implemented, but I was thinking I'd put that in a new PR to avoid leaving this one outstanding too long. I'm guessing there will be some back and forth over whether/how to connect this up with the Euler BCs, how to deal with BCs for `q`, etc.

Reply:
Fair.